sygnal/firebasepushkin.py

```diff
@@ -13,259 +13,419 @@
+                        "Temporary failure, will retry in %d seconds",
+                        retry_delay,
+                        exc_info=True,
+                    )
+                    span_parent.log_kv(
+                        {"event": "temporary_fail", "retrying_in": retry_delay}
+                    )
+                    if retry_number == self.MAX_TRIES - 1:
+                        raise NotificationDispatchException(
+                            "Retried too many times."
+                        ) from ex
+                    else:
+                        await twisted_sleep(
+                            retry_delay, twisted_reactor=self.sygnal.reactor
+                        )
 
-        failed = []
+    @staticmethod
+    def _map_counts_unread(n):
+        return n.counts.unread or 0
 
-        return failed
+    @staticmethod
+    def _map_android_priority(n):
+        """
+        Maps the notification priority coming from synapse to
+        an fcm conform value
+        Args:
+            n: Incoming notification
+
+        Returns:
+            str: Fcm conform priority value
+        """
+        return "normal" if n.prio == "low" else "high"
 
     @staticmethod
-    def build_message(n):
-        data = {}
-        for attribute in [
-            "event_id",
-            "type",
-            "sender",
-            "room_name",
-            "room_alias",
-            "membership",
-            "sender_display_name",
-            "content",
-            "room_id",
-        ]:
-            if hasattr(n, attribute):
-                data[attribute] = getattr(n, attribute)
-                # Truncate fields to a sensible maximum length. If the whole
-                # body is too long, GCM will reject it.
-                if (
-                    data[attribute] is not None
-                    and len(data[attribute]) > MAX_BYTES_PER_FIELD
-                ):
-                    data[attribute] = data[attribute][0:MAX_BYTES_PER_FIELD]
-
-        data["prio"] = "high"
-        if n.prio == "low":
-            data["prio"] = "normal"
-
-        if getattr(n, "counts", None):
-            data["unread"] = n.counts.unread
-            data["missed_calls"] = n.counts.missed_calls
+    def _map_ios_priority(n):
+        """
+        Maps the notification priority coming from synapse to
+        an apns conform value
+        Args:
+            n: Incoming notification
+
+        Returns:
+            int: Apns conform priority value
+        """
+        return "10" if n.prio == "high" else "5"
 
-        return data
+    @staticmethod
+    def _message_notification_content(n, message_types):
+        """
+        Generates the title and body of the visible notification given an
+        incoming notification [n] and [message_types].
+
+        Args:
+            n: The incoming notification
+            message_types (dict of str: str): Mapping from message types to
+            body replacement strings
```

Comment:
```suggestion
                body replacement strings
```

Indenting the overflow lines for arg explanations can improve readability.